llsd/serde_xml.py

```diff
@@ -126,7 +119,46 @@ def _write(self, something):
         """
         self.stream.write(b'<?xml version="1.0" ?>'
                           b'<llsd>')
-        self._generate(something)
+
+        iter_stack = [(iter([something]), b"", None)]
+        while True:
+            cur_iter, iter_type, iterable_obj = iter_stack[-1]
+            try:
+                item = next(cur_iter)
+                if iter_type == b"map":
+
+                    if self.py2: # pragma: no cover
+                        self.stream.write(b'<key>' +
+                                          _str_to_bytes(xml_esc(UnicodeType(item))) +
+                                          b'</key>')
+                    else:
+                        # fair performance improvement by explicitly doing the
+                        # translate for py3 instead of calling xml_esc
+                        self.stream.write(b'<key>' +
+                        UnicodeType(item).translate(XML_ESC_TRANS).encode('utf-8') +
+                                          b'</key>')
+                    item = iterable_obj[item]
+                if isinstance(item, _LLSD):
+                    item = item.thing
+                item_type = type(item)
+                if not item_type in self.type_map:
+                    raise LLSDSerializationError(
+                        "Cannot serialize unknown type: %s (%s)" % (item_type, item))
+                tfunction = self.type_map[item_type]
+
+                if tfunction == self._MAP:
+                    self.stream.write(b'<map>')
+                    iter_stack.append((iter(list(item)), b"map", item))
+                elif tfunction == self._ARRAY:
+                    self.stream.write(b'<array>')
+                    iter_stack.append((iter(item), b"array", None))
+                else:
+                    self.stream.write(tfunction(item))
```

Comment:
It's tempting to suggest that:
* `iter_stack` become an instance attribute
* each of the other type-specific methods actually include the `self.write()` call
* the `if` ... `elif` code in lines 149-154 migrate to the `_MAP()` and `_ARRAY` methods.

Then lines 149-156 become simply `tfunction(item)`.

Reply:
done in future versions.